### PR TITLE
[ADD] account: allow to use a different rate than default one

### PR DESCRIPTION
**Pull request by jjscarafia**
Description of the issue/feature this PR addresses:
Allow to use different rates than the one in odoo when encoding invoices

Current behavior before PR:
When encoding vendor/customer invoices, you're not able to use a different rate than the one on odoo. You neither know the rate after/before invoice validation

Desired behavior after PR is merged:
The improvements after this PR are:
1. User can preview the rate that's going to be used before validating the invoice
2. User can modify the rate before validating invoice
3. You can have information regarding the rate used on an invoice (getting this value calculating  aml.balance / aml.currency_amount or similar is not accurate)

Check [this demo video](https://drive.google.com/file/d/1J28gaUpaCKU1lSsZh4s4w9Wgxema8DsQ/view)

Some remarks:

1. there is an error when changing the rate that creates a new line. This error is already there in odoo and is not related to this PR. Check [this issue](https://github.com/odoo/odoo/issues/35198)
2. This behavior could be optional by some setting like to enable on purchase and or sales

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_move.py [hunk 1]
```diff
@@ -76,6 +76,8 @@ def _get_default_invoice_incoterm(self):
 
     # ==== Business fields ====
     name = fields.Char(string='Number', required=True, readonly=True, copy=False, default='/')
+    rate = fields.Float(digits=(12, 6), help='The between this currency and company currency')
```
Comment (nhomar) [inline addons/account/models/account_move.py]: the rounding should be from decimal_precision engine.

File: addons/account/models/account_move.py [hunk 2]
```diff
@@ -76,6 +76,8 @@ def _get_default_invoice_incoterm(self):
...
+    has_foreign_currency = fields.Boolean(compute='_compute_has_foreign_currency')
```
Comment (nhomar) [inline addons/account/models/account_move.py]: This boolean is not necessary, and in any case this should be store=True to use it in reports.

The account move with empty currency is company currency, any other is foreign.

File: addons/account/models/account_move.py (hunk 1 above)
Comment (jjscarafia) [reply addons/account/models/account_move.py]: I've used the same rounding as in odoo [currency model](https://github.com/odoo/odoo/blob/250b1d3c3f976331743b10202cf30754cc396e6a/odoo/addons/base/models/res_currency.py#L238)

File: addons/account/models/account_move.py (hunk 2 above)
Comment (jjscarafia) [reply addons/account/models/account_move.py]: The computed boolean is there to hide/show the rate field. That was the only purpose, that's why is not stored.

**Comment by jjscarafia** [comment]
Thanks for the review @nhomar.
I've added the demo video.

We also have a custom module for doing this on payments. On payments is really simpler because you don't have computed fields that show amounts on other currency, it's more how to build de journal item after payment post. But of course it would be great to make it odoo core.

On our case, this requirement is mandatory on bills because, for eg:
* a credit note must have the same rate as the original invoice
* Inside Argentina is really common to sell between Argentinian companies in USD. It's mandatory to tell the rate you've use one validating it (so we need the exact rate that was used). Also, when encoding vendor bills, different vendors can use different rates for the same rate, and is mandatory to use that rate (if not VAT may vary between what vendor has and customer has)

I completely agree that is "an oversimplification of the accounting needs of Rate per document". But well, it's a start..

Once more, thanks for the review and feedback!

File: addons/account/models/account_move.py (hunk 2 above)
Comment (nhomar) [reply addons/account/models/account_move.py]: Ok, on that case, I think that you can do a `attrs={'invisible': [('company_id.currency_id.id', '!=', currency_id.id)]}` don't you think? and not new field required at all.

**Comment by hbto** [comment]
Hello @jjscarafia  

> When encoding vendor/customer invoices, you're not able to use a different rate than the one on odoo. 

I kind of agree with you in this statement.

> You neither know the rate after/before invoice validation

I do not agree with you in here.

Let us check this video:  https://youtu.be/1EThxQL6zus

In my opinion it is better to have the absolute values in both currencies, i.e, debit/credit and amount_currency, currency_rates tends to need a lot of decimal points that can always lead to rounding issues. Even setting a rate will lead to not reflect what you are actually trying to convey on your debits and credits and/your amount_currency fields.

That's why it is better to have the rate implicit in the debit/credit - amount_currency values.

**Comment by jjscarafia** [comment]
Hi @hbto . Thanks for the feedback. I agree that we should keep absolute values ("amount_currency" and "debit/credit") but that information is not enough to know which was the rate used. Check this example:
* Company in ARS
* precision of ARS and BRL is 0.01 (typical)
* 11,46 ARS =  1 BRL (real rate between Argentinian currency and Brazilian)
* Invoice of 1.21 BRL --> 0,105584642233857 ARS that will be rounded as 0,11
* If I want to get the rate from the account.move.line (or account.move) then I would get 1,21 / 0,11 = 11

11 is not the rate, it's 11,46. Getting the rate from debit / amount_currency can be inaccurate on many use cases regarding amounts and exchange rates. When we make an electronic invoice we tell the government which rate we have use (11,46 in this case), we need to keep that information for invoice report, for tax declaration, for future refunds, etc.

We don't loose anything, we just keep more information about the exact rate that odoo has used for the conversion. The rate is informative

**Comment by levkar** [comment]
Definetely a good start! Many localizations need this and more... Fixed rate and currency rate type (buy/sell etc.) selection per document is the next.

Thanks for the PR.

File: addons/account/models/account_move.py (hunk 1 above)
Comment (levkar) [reply addons/account/models/account_move.py]: There has to be a "Currency Rate" precision which is configurable. It is missing in the core and we are currently handling it with homegrown currency_rate_precision module.

**Comment by qdp-odoo** [comment]
/me takes good note of this PR

**Comment by bhaveshselarka** [comment]
@jjscarafia

**Comment by jjscarafia** [comment]
yes @bhaveshselarka, how can I help?

**Comment by pedrobaeza** [comment]
Coming to this through codetriage.com. Is this still relevant?

**Comment by jjscarafia** [comment]
Hi @pedrobaeza . Thanks for pinging. Yes it's really important in Argentina, it's mandatory.
I believe that odoo was going to work on something similar. Any news on this @qdp-odoo or @jco-odoo ?

**Comment by jco-odoo** [comment]
We are working on it, but it will not be for the next saas.

**Pull request closed without merging**